File: 10 - TRS using Bates model/Python code/TRS using Bates model.py
# -*- coding: utf-8 -*-
"""
Created on Sat Apr 29 11:13:03 2023

@author: pierre moureaux
"""

import numpy as np
from matplotlib import pyplot as plt

"Market datas"
S0 = 100
sigma0 = 0.1
r = 0.05

"TRS datas"
T = 1
TRSpaymentSchedule = (25,50,75,99)
sTRS = 0.01

"Model datas"
kappa = 0.1
theta = 0.1
gamma = 0.2
rhoSsigma = 0.1
lamb = 20
muJ = 0.1
sigmaJ = 0.1
deltaT = 0.01
lT = int(T/deltaT)
nbSimul = 1000

def pathsGeneration(nbSimul,lT,S0,sigma0,r,kappa,theta,gamma,lamb,muJ,sigmaJ,rhoSsigma):
    paths = np.zeros(nbSimul,dtype=dict)
    for i in range(nbSimul):
        phiS=np.random.normal(0, 1, lT)
        phitild=np.random.normal(0, 1, lT)
        phiJ=np.random.normal(muJ, sigmaJ, lT)
        P = np.random.poisson(deltaT*lamb, lT)
        S = np.zeros(lT)
        sigma = np.zeros(lT)
        intr = np.zeros(lT)
        S[0]=S0
        sigma[0]=sigma0
        intr[0]=r
        meanJRN = np.exp(muJ+0.5*(sigmaJ**2))-1
        if (2*kappa*theta <= gamma**2):
            kappa = (gamma**2)/(2*theta)+0.1
        for k in range(lT-1):
            sigma[k+1] = kappa*(theta-sigma[k])*deltaT + gamma*np.sqrt(sigma[k])*(rhoSsigma*np.sqrt(deltaT)*phiS[k]+np.sqrt(1-rhoSsigma*rhoSsigma)*np.sqrt(deltaT)*phitild[k])+sigma[k]
            S[k+1] = S[k]*((r-lamb*meanJRN)*deltaT+np.sqrt(sigma[k]*deltaT)*phiS[k]+(np.exp(phiJ[k])-1)*P[k]) + S[k]
            intr[k+1]=intr[k]+r
        paths[i] = {"Asset": S, "Volatility": sigma, "Cum. interest rate": intr}
    plt.plot(paths[1]['Asset'])
    return paths

def alphaBetaTRS(TRSpaydate,prevTRSpaydate, paths, nbSimul):
    alphak = 0
    betak = 0
    for i in range(nbSimul):
        res1 = (paths[i]['Asset'][TRSpaydate]-paths[i]['Asset'][prevTRSpaydate])/(np.exp(paths[i]['Cum. interest rate'][TRSpaydate]))
        res2 = (paths[i]['Asset'][prevTRSpaydate])/(np.exp(paths[i]['Cum. interest rate'][TRSpaydate]))
        alphak += res1
        betak += res2
    alphak /= nbSimul
    betak /= nbSimul
    return [alphak,betak]

paths = pathsGeneration(nbSimul,lT,S0,sigma0,r,kappa,theta,gamma,lamb,muJ,sigmaJ,rhoSsigma)

def TRS(TRSpaymentSchedule, sTRS, paths,nbSimul):
    res = 0
    alpha = 0
    beta = 0
    for s in TRSpaymentSchedule:
        alpha += alphaBetaTRS(s,TRSpaymentSchedule[TRSpaymentSchedule.index(s)-1],paths,nbSimul)[0]
        beta += alphaBetaTRS(s,TRSpaymentSchedule[TRSpaymentSchedule.index(s)-1],paths,nbSimul)[1]*(s-TRSpaymentSchedule[TRSpaymentSchedule.index(s)-1])
    res = alpha - sTRS*beta
    impliedRate = alpha/beta
    return [res, impliedRate]

print("TRS market value is",TRS(TRSpaymentSchedule, sTRS, paths,nbSimul)[0])
print("TRS implied rate is",TRS(TRSpaymentSchedule, sTRS, paths,nbSimul)[1])

File: 10 - TRS using Bates model/C++ code/TRS using Bates model.cpp
#include <vector>
#include <unordered_map>
#include <random>
#include <math.h>
#include <any>
#include <iostream>

typedef std::vector<std::unordered_map<std::string, std::vector<double>>> typePaths;

//Market datas
auto S0 = 100.0;
auto sigma0 = 0.1;
auto r = 0.05;

//TRS datas
auto T = 1.0;
std::vector<double> TRSpaymentSchedule = { 0,25, 50, 75, 99 };
auto sTRS = 0.01;

//Model datas
auto kappa = 0.1;
auto theta = 0.1;
auto gamma = 0.2;
auto rhoSsigma = 0.1;
auto lamb = 20;
auto muJ = 0.1;
auto sigmaJ = 0.1;
auto deltaT = 0.01;
auto lT = static_cast<int>(T / deltaT);
auto nbSimul = 1000;

typePaths pathsGeneration(int nbSimul, int lT, double S0, double sigma0, double r, double kappa, double theta, double gamma, 
    double lamb, double muJ, double sigmaJ, double rhoSsigma)
{
    typePaths paths;
    for (auto i = 0; i != nbSimul; i++)
    {
        std::vector<double> phiS;
        std::vector<double> phitild;
        std::vector<double> phiJ;
        std::vector<double> P;
        std::default_random_engine generator;
        for (auto k = 0; k != lT; k++)
        {
            phiS.emplace_back(std::normal_distribution<double>{0, 1}(generator));
            phitild.emplace_back(std::normal_distribution<double>{0, 1}(generator));
            phiJ.emplace_back(std::normal_distribution<double>{muJ, sigmaJ}(generator));
            P.emplace_back(std::poisson_distribution<int>{deltaT*lamb}(generator));
        }
        std::vector<double> S(lT);
        std::vector<double> sigma(lT);
        std::vector<double> intr(lT);
        S[0] = S0;
        sigma[0] = sigma0;
        intr[0] = r;
        auto meanJRN = exp(muJ + 0.5 * (sigmaJ*sigmaJ)) - 1;
        if (2 * kappa * theta <= (gamma*gamma))
        {
            kappa = (gamma*gamma) / (2 * theta) + 5;
        }
        for (auto k = 0; k != (lT - 1); k++)
        {
            sigma[k + 1] = kappa * (theta - sigma[k]) * deltaT + gamma * sqrt(sigma[k]) * (rhoSsigma * sqrt(deltaT) * phiS[k] + sqrt(1 - rhoSsigma * rhoSsigma) * sqrt(deltaT) * phitild[k]) + sigma[k];
            S[k + 1] = S[k] * ((r - lamb * meanJRN)* deltaT + sqrt(sigma[k] * deltaT) * phiS[k] + (exp(phiJ[k]) - 1) * P[k]) + S[k];
            intr[k + 1] = intr[k] + r;
        }
        std::unordered_map<std::string, std::vector<double>> interRes = { {"Asset",S},{"Volatility",sigma},{"Cum. interest rate",intr} };
        paths.emplace_back(interRes);
    }
    return paths;
}

std::vector<double> alphaBetaTRS(int TRSpaydate, int prevTRSpaydate, const typePaths& paths, int nbSimul)
{
    auto alphak = 0.0;
    auto betak = 0.0;
    for (auto i = 0; i != nbSimul; i++)
    {
        auto res1 = (paths[i].at("Asset")[TRSpaydate] - paths[i].at("Asset")[prevTRSpaydate]) / (exp(paths[i].at("Cum. interest rate")[TRSpaydate]));
        auto res2 = (paths[i].at("Asset")[prevTRSpaydate]) / (exp(paths[i].at("Cum. interest rate")[TRSpaydate]));
        alphak += res1;
        betak += res2;
    }
    alphak /= nbSimul;
    betak /= nbSimul;
    std::vector<double> res;
    res.emplace_back(alphak);
    res.emplace_back(betak);
    return res;
}

int getIndex(const std::vector<double>& v, int K)
{
    auto it = find(v.begin(), v.end(), K);
    if (it != v.end())
    {
        int index = it - v.begin();
        return index;
    }
    else
    {
        return 1;
    }
}

std::vector<double> TRS(const std::vector<double>& TRSpaymentSchedule, double sTRS, const typePaths& paths, int nbSimul)
{
    auto alpha = 0.0;
    auto beta = 0.0;
    for (auto s = TRSpaymentSchedule.begin(); s != TRSpaymentSchedule.end() - 1; s++)
    {
        auto payDate = getIndex(TRSpaymentSchedule, *s) + 1;
        auto prevPayDate = getIndex(TRSpaymentSchedule, *s);
        auto abRes = alphaBetaTRS(TRSpaymentSchedule[payDate], TRSpaymentSchedule[prevPayDate], paths, nbSimul);
        alpha += abRes[0];
        beta += abRes[1] * (TRSpaymentSchedule[payDate] - TRSpaymentSchedule[prevPayDate]);
    }
    auto resTemp = alpha - sTRS * beta;
    auto impliedRate = alpha / beta;
    std::vector<double> res;
    res.emplace_back(resTemp);
    res.emplace_back(impliedRate);
    return res;
}

int main()
{
    auto paths = pathsGeneration(nbSimul, lT, S0, sigma0, r, kappa, theta, gamma,lamb, muJ, sigmaJ, rhoSsigma);
    auto TRSres = TRS(TRSpaymentSchedule, sTRS, paths, nbSimul);
    std::cout << "TRS market value is" << TRSres[0] << std::endl;
    std::cout << "TRS implied rate is" << TRSres[1] << std::endl;
}
